apisix/core/request.lua

```diff
@@ -143,14 +143,35 @@ function _M.set_header(ctx, header_name, header_value)
     end
 end
 
-function _M.add_header(header_name, header_value)
+function _M.add_header(ctx, header_name, header_value)
+    if type(ctx) == "string" then
+        header_value = header_name
+        header_name = ctx
+        ctx = nil
+
+        log.warn("DEPRECATED: use add_header(ctx, header_name, header_value) instead")
+    end
+
     local err
     header_name, err = _validate_header_name(header_name)
     if err then
         error(err)
     end
 
+    local changed = false
+    if is_apisix_or then
+        changed = a6_request.is_request_header_set()
+    end
+
+
     req_add_header(header_name, header_value)
+
+    if is_apisix_or and not changed then
+        a6_request.clear_request_header()
+        if ctx and ctx.headers then
+            ctx.headers[header_name] = header_value
```

Comment:
We should append instead of overwriting it?

Reply:
Fixed